src/Functions/parseDateTime.cpp

```diff
@@ -0,0 +1,1747 @@
+
+            auto col_res = ColumnDateTime::create();
+            col_res->reserve(input_rows_count);
+            auto & data_res = col_res->getData();
+            Date date;
+            for (size_t i = 0; i < input_rows_count; ++i)
+            {
+                date.reset();
+
+                StringRef str_ref = col_str->getDataAt(i);
+                Pos cur = str_ref.data;
+                Pos end = str_ref.data + str_ref.size;
+                for (const auto & instruction : instructions)
+                {
+                    // std::cout << "instruction:" << instruction.toString() << std::endl;
+                    cur = instruction.perform(cur, end, date);
+                    // std::cout << "date:" << date.toString() << std::endl;
+                }
+
+                // Ensure all input was consumed.
+                if (cur < end)
+                    throw Exception(
+                        ErrorCodes::LOGICAL_ERROR,
+                        "Invalid format input {} is malformed at {}",
+                        str_ref.toView(),
+                        std::string_view(cur, end - cur));
+
+                Int64 time = date.checkAndGetDateTime(*time_zone);
+                data_res.push_back(static_cast<UInt32>(time));
+            }
+
+            return col_res;
+        }
+
+
+    private:
+        class Action
+        {
+        private:
+            using Func = std::conditional_t<
+                parse_syntax == ParseDateTimeTraits::ParseSyntax::MySQL,
+                Pos (*)(Pos, Pos, Date &),
+                std::function<Pos(Pos, Pos, Date &)>>;
+            Func func{};
+            std::string func_name;
+
+            std::string literal;
+
+        public:
+            explicit Action(Func && func_, const char * func_name_) : func(std::move(func_)), func_name(func_name_) { }
+
+            explicit Action(const String & literal_) : literal(literal_) { }
+            explicit Action(String && literal_) : literal(std::move(literal_)) { }
+
+            /// For debug
+            [[maybe_unused]] String toString() const
+            {
+                if (func)
+                    return "func:" + func_name;
+                else
+                    return "literal:" + literal;
+            }
+
+            Pos perform(Pos cur, Pos end, Date & date) const
+            {
+                if (func)
+                    return func(cur, end, date);
+                else
+                {
+                    ensureSpace(cur, end, literal.size(), "required literal size not matched");
+                    if (std::string_view(cur, literal.size()) != literal)
+                        throw Exception(
+                            ErrorCodes::LOGICAL_ERROR, "Expect literal {} but {} provided", literal, std::string_view(cur, literal.size()));
+                    cur += literal.size();
+                    return cur;
+                }
+            }
+
+            template <typename T, bool check_space = true>
```

Comment:
Since the `check_space` template argument is used repeatedly (here, l., 655, l. 668, l. 683): Could we make it an `enum`?

```
enum class CheckSpace
{
    Yes,
    No
};

template < typename T, CheckSpace check_space>
static Pos readNumber2(Pos cur, Pos end, T & res)
```

Invocation of `readNumber2()` will look a more aesthetic: `readNumber<CheckSpace::Yes>(...)`.

I also don't really like optional template arguments (`=`, l. 654)too much, they require the reader to read the function definition to understand the actual behavior. My preference would be to make `check_space` mandatory.

Another alternative is to remove all space checking from `readNumber*()` and instead simply let the caller do that (e.g. `mySqlMonth()`).